File: packages/lightning/src/bridge/commands.ts
import type { bridge_data } from '../database/mod.ts';
import { bridge_settings_list } from '../structures/bridge.ts';
import { log_error } from '../structures/errors.ts';
import type { bridge_channel, command_opts } from '../structures/mod.ts';

export async function create(
	db: bridge_data,
	opts: command_opts,
): Promise<string> {
	const result = await _add(db, opts);

	if (typeof result === 'string') return result;

	const data = {
		name: opts.args.name!,
		channels: [result],
		settings: {
			allow_editing: true,
			allow_everyone: false,
			use_rawname: false,
		},
	};

	try {
		const { id } = await db.create_bridge(data);
		return `Bridge created successfully!\nYou can now join it using \`${opts.prefix}bridge join ${id}\`.\nKeep this ID safe, don't share it with anyone, and delete this message.`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to insert bridge into database',
			extra: data,
		});
	}
}

export async function join(
	db: bridge_data,
	opts: command_opts,
): Promise<string> {
	const result = await _add(db, opts);

	if (typeof result === 'string') return result;

	const target_bridge = await db.get_bridge_by_id(
		opts.args.id!,
	);

	if (!target_bridge) {
		return `Bridge with id \`${opts.args.id}\` not found. Make sure you have the correct id.`;
	}

	target_bridge.channels.push(result);

	try {
		await db.edit_bridge(target_bridge);

		return `Bridge joined successfully!`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to update bridge in database',
			extra: { target_bridge },
		});
	}
}

async function _add(
	db: bridge_data,
	opts: command_opts,
): Promise<string | bridge_channel> {
	const existing_bridge = await db.get_bridge_by_channel(
		opts.channel_id,
	);

	if (existing_bridge) {
		return `You are already in a bridge called \`${existing_bridge.name}\`. You must leave it before being in another bridge. Try using \`${opts.prefix}bridge leave\` or \`${opts.prefix}help\` commands.`;
	}

	try {
		return {
			id: opts.channel_id,
			data: await opts.plugin.setup_channel(opts.channel_id),
			disabled: false,
			plugin: opts.plugin.name,
		};
	} catch (e) {
		log_error(e, {
			message: 'Failed to create bridge using plugin',
			extra: { channel: opts.channel_id, plugin_name: opts.plugin },
		});
	}
}

export async function leave(
	db: bridge_data,
	opts: command_opts,
): Promise<string> {
	const bridge = await db.get_bridge_by_channel(
		opts.channel_id,
	);

	if (!bridge) return `You are not in a bridge`;

	if (opts.args.id !== bridge.id) {
		return `You must provide the bridge id in order to leave this bridge`;
	}

	bridge.channels = bridge.channels.filter((
		ch,
	) => ch.id !== opts.channel_id);

	try {
		await db.edit_bridge(
			bridge,
		);
		return `Bridge left successfully`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to update bridge in database',
			extra: { bridge },
		});
	}
}

export async function status(
	db: bridge_data,
	opts: command_opts,
): Promise<string> {
	const bridge = await db.get_bridge_by_channel(
		opts.channel_id,
	);

	if (!bridge) return `You are not in a bridge`;

	let str = `Name: \`${bridge.name}\`\n\nChannels:\n`;

	for (const [i, value] of bridge.channels.entries()) {
		str += `${i + 1}. \`${value.id}\` on \`${value.plugin}\`${
			value.disabled ? ' (disabled)' : ''
		}\n`;
	}

	str += `\nSettings:\n`;

	for (
		const [key, value] of Object.entries(bridge.settings).filter(([key]) =>
			bridge_settings_list.includes(key)
		)
	) {
		str += `- \`${key}\` ${value ? '✔' : '❌'}\n`;
	}

	return str;
}

export async function toggle(
	db: bridge_data,
	opts: command_opts,
): Promise<string> {
	const bridge = await db.get_bridge_by_channel(
		opts.channel_id,
	);

	if (!bridge) return `You are not in a bridge`;

	if (!bridge_settings_list.includes(opts.args.setting!)) {
		return `That setting does not exist`;
	}

	const key = opts.args.setting as keyof typeof bridge.settings;

	bridge.settings[key] = !bridge.settings[key];

	try {
		await db.edit_bridge(
			bridge,
		);
		return `Bridge settings updated successfully`;
	} catch (e) {
		log_error(e, {
			message: 'Failed to update bridge in database',
			extra: { bridge },
		});
	}
}

File: internal/bridge/commands.go
package bridge

import (
	"slices"
	"strconv"

	"github.com/oklog/ulid/v2"
	"github.com/williamhorning/lightning/pkg/lightning"
)

const notInBridge = "You're not in a bridge"

func bridgeCommand(database Database) lightning.Command {
	return lightning.Command{
		Name:        "bridge",
		Description: "manage bridges between channels",
		Executor: func(_ lightning.CommandOptions) (string, error) {
			return "take a look at the subcommands of this command", nil
		},
		Subcommands: []lightning.Command{
			{
				Name:        "create",
				Description: "create a new bridge",
				Arguments: []lightning.CommandArgument{
					{Name: "name", Description: "the name to use for the bridge", Required: true},
				},
				Executor: func(options lightning.CommandOptions) (string, error) {
					return createCommand(database, options)
				},
			},
			{
				Name:        "join",
				Description: "join an existing bridge",
				Arguments:   arguments("join"),
				Executor: func(options lightning.CommandOptions) (string, error) {
					return joinCommand(database, options, false)
				},
			},
			{
				Name:        "subscribe",
				Description: "subscribe to a bridge",
				Arguments:   arguments("subscribe to"),
				Executor: func(options lightning.CommandOptions) (string, error) {
					return joinCommand(database, options, true)
				},
			},
			{
				Name:        "leave",
				Description: "leave a bridge",
				Arguments:   arguments("leave"),
				Executor: func(options lightning.CommandOptions) (string, error) {
					return leaveCommand(database, options)
				},
			},
			{
				Name:        "toggle",
				Description: "toggle settings for a bridge",
				Arguments: []lightning.CommandArgument{
					{Name: "setting", Description: "the bridge setting to toggle", Required: true},
				},
				Executor: func(options lightning.CommandOptions) (string, error) {
					return toggleCommand(database, options)
				},
			},
			{
				Name:        "status",
				Description: "get the status of the bridge in this channel",
				Executor: func(options lightning.CommandOptions) (string, error) {
					return statusCommand(database, options)
				},
			},
		},
	}
}

func arguments(to string) []lightning.CommandArgument {
	return []lightning.CommandArgument{
		{Name: "id", Description: "the ID of the bridge to " + to, Required: true},
	}
}

func prepareChannelForBridge(db Database, opts lightning.CommandOptions) (bridgeChannel, string) {
	if br, err := getBridgeByChannel(db, opts.ChannelID); br.ID != "" || err != nil {
		return bridgeChannel{}, "This channel is already part of a bridge. Please leave the bridge first."
	}

	data, err := opts.Bot.SetupChannel(opts.ChannelID)
	if err != nil {
		return bridgeChannel{}, lightning.LogError(err, "Failed to setup channel for bridge",
			map[string]any{"channel": opts.ChannelID}, nil).Error()
	}

	return bridgeChannel{
		ID:       opts.ChannelID,
		Data:     data,
		Disabled: lightning.ChannelDisabled{},
	}, ""
}

func createCommand(database Database, opts lightning.CommandOptions) (string, error) {
	channel, errMsg := prepareChannelForBridge(database, opts)
	if errMsg != "" {
		return errMsg, nil
	}

	bridgeData := bridge{
		ID:       ulid.Make().String(),
		Name:     opts.Arguments["name"],
		Channels: []bridgeChannel{channel},
		Settings: bridgeSettings{false},
	}

	if err := database.createBridge(bridgeData); err != nil {
		return lightning.LogError(err, "Failed to create bridge in database",
			map[string]any{"bridge": bridgeData}, nil).Error(), nil
	}

	join := opts.Prefix + "bridge join " + bridgeData.ID

	return "Bridge created successfully! You can now join it using ||`" + join + "`||. Keep this command secret!", nil
}

func joinCommand(database Database, opts lightning.CommandOptions, subscribe bool) (string, error) {
	channel, errMsg := prepareChannelForBridge(database, opts)
	if errMsg != "" {
		return errMsg, nil
	}

	bridgeData, err := database.getBridge(opts.Arguments["id"])
	if err != nil {
		return lightning.LogError(err, "Failed to get bridge from database",
			map[string]any{"bridge_id": opts.Arguments["id"]}, nil).Error(), nil
	} else if bridgeData.ID == "" {
		return "No bridge found with the provided ID.", nil
	}

	channel.Disabled = lightning.ChannelDisabled{Read: subscribe, Write: false}
	bridgeData.Channels = append(bridgeData.Channels, channel)

	if err := database.createBridge(bridgeData); err != nil {
		return lightning.LogError(err, "Failed to update bridge in database",
			map[string]any{"bridge": bridgeData}, nil).Error(), nil
	}

	return "Bridge joined successfully!", nil
}

func leaveCommand(database Database, opts lightning.CommandOptions) (string, error) {
	bridgeData, err := getBridgeByChannel(database, opts.ChannelID)
	if err != nil {
		return lightning.LogError(err, "Failed to get bridge from database",
			map[string]any{"channel": opts.ChannelID}, nil).Error(), nil
	} else if bridgeData.ID == "" {
		return notInBridge, nil
	}

	if bridgeData.ID != opts.Arguments["id"] {
		return "This channel is not part of the specified bridge.", nil
	}

	for idx, channel := range bridgeData.Channels {
		if compareChannelIDs(channel, opts.ChannelID) {
			bridgeData.Channels = slices.Delete(bridgeData.Channels, idx, idx+1)

			break
		}
	}

	if err := database.createBridge(bridgeData); err != nil {
		return lightning.LogError(err, "Failed to update bridge in database",
			map[string]any{"bridge": bridgeData}, nil).Error(), nil
	}

	return "You have successfully left the bridge.", nil
}

func toggleCommand(database Database, opts lightning.CommandOptions) (string, error) {
	setting := opts.Arguments["setting"]

	bridgeData, err := getBridgeByChannel(database, opts.ChannelID)
	if err != nil {
		return lightning.LogError(err, "Failed to get bridge from database",
			map[string]any{"channel": opts.ChannelID}, nil).Error(), nil
	} else if bridgeData.ID == "" {
		return notInBridge, nil
	}

	if setting != "allow_everyone" {
		return "That setting does not exist. Available settings are: `allow_everyone`.", nil
	}

	bridgeData.Settings.AllowEveryone = !bridgeData.Settings.AllowEveryone

	if err := database.createBridge(bridgeData); err != nil {
		return lightning.LogError(err, "Failed to update bridge in database",
			map[string]any{"bridge": bridgeData}, nil).Error(), nil
	}

	return "Bridge settings updated successfully", nil
}

func statusCommand(db Database, opts lightning.CommandOptions) (string, error) {
	bridgeData, err := getBridgeByChannel(db, opts.ChannelID)
	if err != nil {
		return lightning.LogError(err, "Failed to get bridge from database",
			map[string]any{"channel": opts.ChannelID}, nil).Error(), nil
	} else if bridgeData.ID == "" {
		return notInBridge, nil
	}

	status := "Name: `" + bridgeData.Name + "`\n\nChannels:\n"

	for i, channel := range bridgeData.Channels {
		status += strconv.Itoa(i+1) + ". `"

		if channel.DeprecatedPlugin == "" {
			status += channel.ID + "`"
		} else {
			status += channel.DeprecatedPlugin + "::" + channel.ID + "`"
		}

		if channel.isDisabled().Read {
			status += " (subscribed)"
		}

		if channel.isDisabled().Write {
			status += " (write disabled)"
		}

		status += "\n"
	}

	status += "\nSettings:\n"
	status += "- AllowEveryone: `" + (map[bool]string{true: "✔", false: "❌"})[bridgeData.Settings.AllowEveryone] + "`\n"

	return status, nil
}
